### PR TITLE
Adds WFS support

### PR DESCRIPTION
**Pull request by tkohr**
This PR enables printing WFS on map using default OL style. Changing opacity is not part of the PR as it belongs to the style.

The spec allows to define the WFS `version` (1.0.0, 1.1.0, 2.0.0) and data `format` (geojson, gml) to use. The GML version used, is currently coupled to the WFS version indicated. (There is still a problem with WFS 2.0.0)

The PR does not monitor the download progress, as all data is loaded at once.

TODOS:

- [x] make WFS 2.0.0 (with GML) work
- [x] make rendering test work with local resource
- [x] add unit test for url generation

### CODE REVIEW

File: src/printer/layers.js [hunk 1]
```diff
@@ -267,3 +277,120 @@ function createLayerWMTS(layerSpec, rootFrameState) {
...
+  let vectorSource = new VectorSource({
+    format: format,
+    loader: function (extent, resolution, projection) {
+      let proj = projection.getCode();
+      let xhr = new XMLHttpRequest();
+      let params = {
+        service: 'WFS',
+        request: 'GetFeature',
+        version: layerSpec.version ? layerSpec.version : '1.1.0',
+        typename: layerSpec.layer,
```
Comment (fgravin) [inline src/printer/layers.js:318]: not the same name for WFS 2.0.0

File: src/printer/layers.js [hunk 2]
```diff
@@ -267,3 +277,120 @@ function createLayerWMTS(layerSpec, rootFrameState) {
...
+        srsName: proj,
+        bbox: extent.join(','),
+      };
+      if (layerSpec.format !== 'gml') {
+        params['outputFormat'] = 'application/json';
+      }
+      let urlParams = Object.entries(params)
```
Comment (fgravin) [inline src/printer/layers.js:325]: Use something like this to build the url https://github.com/camptocamp/mtes-carto-ui/blob/81606d6860dd7dc7f8dcb153c176186307b9dc8d/libs/feature/shared/src/lib/helpers/ogc.helpers.ts#L17

File: src/printer/layers.js [hunk 3]
```diff
@@ -267,3 +277,120 @@ function createLayerWMTS(layerSpec, rootFrameState) {
...
+        .map((e) => e.join('='))
+        .join('&');
+      let url = layerSpec.url + '?' + urlParams;
+      xhr.open('GET', url);
+      let onError = function () {
+        vectorSource.removeLoadedExtent(extent);
```
Comment (fgravin) [inline src/printer/layers.js:331]: Do you raise the error to the client ?

File: src/printer/layers.js [hunk 4]
```diff
@@ -267,3 +277,120 @@ function createLayerWMTS(layerSpec, rootFrameState) {
...
+      xhr.send();
+    },
+    strategy: bbox,
+  });
+
+  let layer = new VectorLayer({
+    source: vectorSource,
+  });
+
+  frameState = {
```
Comment (fgravin) [inline src/printer/layers.js:361]: can this be factorized ?

File: test/unit/printer/layers.test.js [hunk 5]
```diff
@@ -190,6 +190,32 @@ describe('layer creation', () => {
...
+      completed = false;
+      layer$ = createLayer(spec, { ...frameState });
+      layer$.subscribe(
+        (status) => (received = status),
+        null,
+        () => (completed = true)
+      );
+    });
+
+    it('initially emit a status with progress 0', () => {
```
Comment (fgravin) [inline test/unit/printer/layers.test.js:213]: can't you test the url of the WFS GetFeatures ?

File: src/printer/layers.js [hunk 6]
```diff
@@ -267,3 +277,120 @@ function createLayerWMTS(layerSpec, rootFrameState) {
...
+  };
+
+  renderer = layer.getRenderer();
+  renderer.useContainer = function () {
+    this.containerReused = false;
+    this.canvas = context.canvas;
+    this.context = context;
+    this.container = {
+      firstElementChild: context.canvas,
+      style: {
```
Comment (fgravin) [inline src/printer/layers.js:386]: is it mandatory as the opacity is in the layer state already ?

File: src/printer/layers.js [hunk 7]
```diff
@@ -267,3 +277,120 @@ function createLayerWMTS(layerSpec, rootFrameState) {
...
+        opacity: 1,
+        sourceState: 'ready',
+        visible: true,
+        zIndex: 0,
+      },
+    ],
+  };
+
+  renderer = layer.getRenderer();
+  renderer.useContainer = function () {
```
Comment (fgravin) [inline src/printer/layers.js:380]: same ?

File: src/printer/layers.js [hunk 8]
```diff
@@ -267,3 +277,120 @@ function createLayerWMTS(layerSpec, rootFrameState) {
...
+      gmlFormat: gmlFormats[layerSpec.version],
+    });
+  } else {
+    format = new GeoJSON();
+  }
+
+  let vectorSource = new VectorSource({
+    format: format,
+    loader: function (extent, resolution, projection) {
+      let proj = projection.getCode();
```
Comment (fgravin) [inline src/printer/layers.js:312]:
```suggestion
      const projCode = projection.getCode();
```

File: src/printer/layers.js [hunk 9]
```diff
@@ -267,3 +277,120 @@ function createLayerWMTS(layerSpec, rootFrameState) {
...
+      let xhr = new XMLHttpRequest();
```
Comment (fgravin) [inline src/printer/layers.js:313]:
```suggestion
      const xhr = new XMLHttpRequest();
```

**Review by fgravin — COMMENTED** [review]
Very good starts, it works well.
Some comments about code optimization.
Tests are a bit light

File: src/printer/layers.js (hunk 1 above)
Comment (tkohr) [reply src/printer/layers.js]: Ah that's right, thanks. Somehow WFS 2.0.0 still doesn't work. When I try to open the `GetFeature` url in the browser it downloads the response for 2.0.0. Why? Could this be the reason?

File: src/printer/layers.js (hunk 2 above)
Comment (tkohr) [reply src/printer/layers.js]: Thanks, I adapted

File: src/printer/layers.js (hunk 3 above)
Comment (tkohr) [reply src/printer/layers.js:341]: Yes, after rebase on merged errorhandling the url is raised into the job status

File: src/printer/layers.js (hunk 4 above)
Comment (tkohr) [reply src/printer/layers.js:371]: Indeed I have the impression that we only need the `layer` in the `layerStatesArray` for vectorlayers, but I don't think I have understood this part entirely.

File: test/unit/printer/layers.test.js (hunk 5 above)
Comment (tkohr) [reply test/unit/printer/layers.test.js:261]: Hum, good point. How do I trigger the `VectorSource.loader()` (to construct the url).

File: src/printer/layers.js (hunk 6 above)
Comment (tkohr) [reply src/printer/layers.js:387]: If I dont' indicate it here I get an error from ol `VectorLayer`

File: src/printer/layers.js (hunk 7 above)
Comment (tkohr) [reply src/printer/layers.js:381]: I don't understand

Comment (fgravin) [reply src/printer/layers.js:381]: Can it be factorized ? Seems it's almost the same for all layers isn't it ?

Comment (tkohr) [reply src/printer/layers.js:381]: Ah yes of course

File: test/unit/printer/layers.test.js (hunk 5 above)
Comment (fgravin) [reply test/unit/printer/layers.test.js:261]: I don't know :)

**Comment by tkohr** [comment]
Thanks for your fix @fgravin. Nice that it also works without forcing GML3.

I added a test for url generation and moved rendering test to local resource. Saw CORS issue popping up a couple of times, which disappeared again by itself. From my side PR is ready.

**Pull request merged**